tds/src/main/java/thredds/servlet/ServletUtil.java

```diff
@@ -247,6 +229,96 @@ public static void returnFile(HttpServletRequest req, HttpServletResponse res, F
     }
   }
 
+  private static void addContentLengthHeader(HttpServletResponse res, long contentLength) {
+    // when compression is turned on, ContentLength has to be overridden
+    // this is also true for HEAD, since this must be the same as GET without the body
+    if (contentLength > Integer.MAX_VALUE)
+      res.addHeader("Content-Length", Long.toString(contentLength)); // allow content length > MAX_INT
+    else
+      res.setContentLength((int) contentLength);
+  }
+
+  private static boolean isRangeRequest(String rangeRequest) {
+    if (rangeRequest != null) { // bytes=12-34 or bytes=12-
+      return rangeRequest.indexOf("=") > 0 && rangeRequest.indexOf("-") > 0;
+    }
+
+    return false;
+  }
+
+  private static long getContentStartPosition(String rangeRequest) {
+    final boolean isRangeRequest = isRangeRequest(rangeRequest);
+
+    if (!isRangeRequest)
+      return 0;
+
+    // bytes=12-34 or bytes=12-
+    final String startString = rangeRequest.substring(rangeRequest.indexOf("=") + 1, rangeRequest.indexOf("-"));
+    return Long.parseLong(startString);
+  }
+
+  private static long getContentEndPosition(String rangeRequest, Long fileLength) {
+    final boolean isRangeRequest = isRangeRequest(rangeRequest);
+
+    if (!isRangeRequest)
+      return fileLength;
+
+    // bytes=12-34 or bytes=12-
+    final String endString = rangeRequest.substring(rangeRequest.indexOf("-") + 1);
+    final long endPosition = endString.length() > 0 ? Long.parseLong(endString) + 1 : fileLength;
+    return Math.min(endPosition, fileLength);
+  }
+
+  /**
+   * Write an MFile to the response stream.
+   *
+   * @param request the HttpServletRequest
+   * @param response the HttpServletResponse
+   * @param requestPath the request path
+   * @throws IOException if an I/O error occurs while writing the response.
+   */
+  public static void writeMFileToResponse(HttpServletRequest request, HttpServletResponse response, String requestPath)
+      throws IOException {
+    final String location = TdsRequestedDataset.getLocationFromRequestPath(requestPath);
+    final MFile file = MFiles.create(location);
+
+    if (file == null) {
+      response.sendError(HttpServletResponse.SC_NOT_FOUND, "Could not find file: " + location);
+      return;
+    }
+
+    if (file.isDirectory()) {
```

Comment:
Just a note that at some point we might want to let this return a zipped directory